pkg/virtual/initializingworkspaces/builder/caching_authorizer.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package builder
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	kcpkubernetesclientset "github.com/kcp-dev/client-go/kubernetes"
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/apimachinery/pkg/util/cache"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/pkg/apis/tenancy/initialization"
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/authorization/delegated"
+	dynamiccontext "github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/context"
+)
+
+func newCachingAuthorizer(client kcpkubernetesclientset.ClusterInterface) *cachingAuthorizer {
+	return &cachingAuthorizer{client: client, cache: cache.NewExpiring()}
+}
+
+// cachingAuthorizer is a wrapper around athorizer.Authorize that uses
+// an internal expiring cache.
+type cachingAuthorizer struct {
+	client kcpkubernetesclientset.ClusterInterface
+	cache  *cache.Expiring
+}
+
+// load loads the authorizer from the cache, if any.
+func (c *cachingAuthorizer) load(clusterName logicalcluster.Name) authorizer.Authorizer {
+	value, ok := c.cache.Get(clusterName)
+	if !ok && value == nil {
+		return nil
+	}
+	authz, ok := value.(authorizer.Authorizer)
+	if !ok {
+		return nil
+	}
+	return authz
+}
+
+func (c *cachingAuthorizer) loadOrStore(clusterName logicalcluster.Name) (authorizer.Authorizer, error) {
+	if authz := c.load(clusterName); authz != nil {
+		return authz, nil
+	}
+
+	// Create the delegated authorizer.
+	authz, err := delegated.NewDelegatedAuthorizer(clusterName, c.client)
+	if err != nil {
+		return nil, err
+	}
+
+	// Store the cache and return
+	c.cache.Set(clusterName, authz, 12*time.Hour)
+	return authz, nil
+}
+
+func (c *cachingAuthorizer) Authorize(ctx context.Context, attr authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+	logger := klog.FromContext(ctx)
+	workspace, name, err := initialization.TypeFrom(tenancyv1alpha1.ClusterWorkspaceInitializer(dynamiccontext.APIDomainKeyFrom(ctx)))
+	if err != nil {
+		logger.V(2).Info(err.Error())
+		return authorizer.DecisionNoOpinion, "unable to determine initializer", fmt.Errorf("access not permitted")
```

Comment:
@s-urbaniak PTAL

Reply:
This is taken 1:1 from the current implementation, so if anything I suggest to take improvements as a separate PR. However, my comment for this (and thus the previous implementation):

- specifying a reason in case of errors is not necessary. If an error occurs, the reason is discarded as per https://github.com/kubernetes/kubernetes/blob/00aae4c10c204d258e62a96f30970e9b91b29149/staging/src/k8s.io/apiserver/pkg/endpoints/filters/authorization.go#L67-L71
- We have to be careful of anonymizing here, so using the authorization decorator https://github.com/kcp-dev/kcp/blob/ffc21ea3df5d5ce692d9a05665c690fd483ebe95/pkg/authorization/decorator.go#L48 to anonymize is the way to go.

However, it's ok to address this as a follow-up as this PR introduces caching of delegated authorizers only.